client/client.go

```diff
@@ -0,0 +1,391 @@
+package client
+
+import (
+	"errors"
+	"net"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/eclipse/paho.mqtt.golang/packets"
+	"github.com/mainflux/fluxmq/auth"
+	"github.com/mainflux/fluxmq/topic"
+	"go.uber.org/zap"
+)
+
+const (
+	qos0 byte = iota
+	qos1
+	qos2
+	qosFail = 0x80
+)
+
+var groupRegexp = regexp.MustCompile(`^\$share/([0-9a-zA-Z_-]+)/(.*)$`)
+
+// ClientInfo stores basic MQTT client data.
+// Separated in the struct so that it can be
+// easily used as a param of Auth handler
+type ClientInfo struct {
+	ID       string
+	Username string
+	Password []byte
+}
+
+// Client stores complete MQTT client data
+type Client struct {
+	info      ClientInfo
+	conn      net.Conn
+	keepalive uint16
+	lwt       *packets.PublishPacket
+	handler   auth.Handler
+	subs      map[string]topic.Subsciption
+	logger    *zap.Logger
+}
+
+func New(info ClientInfo, keepalive uint16, lwt *packets.PublishPacket, handler auth.Handler, logger *zap.Logger) Client {
+	return Client{
+		info:      info,
+		keepalive: keepalive,
+		lwt:       lwt,
+		handler:   handler,
+		subs:      make(map[string]topic.Subsciption),
+		logger:    logger,
+	}
+}
+
+func NewInfo(clientID, username string, password []byte) ClientInfo {
+	return ClientInfo{
+		ID:       clientID,
+		Username: username,
+		Password: password,
+	}
+}
+
+func (c Client) ReadLoop() {
+	keepAlive := time.Second * time.Duration(c.info.keepalive)
+	timeOut := keepAlive + (keepAlive / 2)
+	dpkt := packets.NewControlPacket(packets.Disconnect).(*packets.DisconnectPacket)
+
+	for {
+		select {
+		case <-c.ctx.Done():
```

Comment:
`ctx`? I guess you'll add it later

Reply:
For now this is just a copy of some other code, I need to clean... I posted PR just for early preview.